### PR TITLE
Rchain 2365 unbond and wait

### PR DESCRIPTION
**Pull request by zsluedem**
## Overview
1. add getActiveValidator api to runtimeManager
2. check the node if it is an active validator to propose
3. add the unbond intergration tests
4. add bonding integrations tests

### JIRA ticket:
https://rchain.atlassian.net/browse/RCHAIN-2365

### Notes
<sup>_Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else._</sup>

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

**Comment by zsluedem** [comment]
bors try

File: integration-tests/test/test_unbond.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        conftest.temporary_wallets_file(random_generator) as wallets_file, \
+        started_bootstrap(context=context, network=network,mount_dir=context.mount_dir, wallets_file=wallets_file) as bootstrap_node, \
+        bootstrap_connected_peer(context=context, bootstrap=bootstrap_node, name='bonded-validator-1', private_key=VALIDATOR_KEY_1) as validator_1:
+
+        validator_1.deploy('/opt/docker/examples/tut-hello.rho', VALIDATOR_KEY_1)
+        validator_1.propose()
+
+        block_hash = validator_1.deploy_contract_with_substitution(substitute_dict={}, rho_file_path="resources/wallets/unbond.rho", private_key=VALIDATOR_KEY_1)
+
+        wait_for_node_sees_block(context, bootstrap_node, block_hash)
```
Comment (tgrospic) [inline integration-tests/test/test_unbond.py]: This waiting uses shell to start a new instance of node to make gRPC request to checking node.
https://github.com/rchain/rchain/blob/5351708/integration-tests/test/rnode.py#L263

This seems super slow. Can we use gRPC directly or events API to check for a block? Although this is probably a change for another ticket to get rid of shell calls in rnode.py.

**Review by tgrospic — APPROVED** [review]

File: integration-tests/test/test_unbond.py (hunk 1 above)
Comment (zsluedem) [reply integration-tests/test/test_unbond.py]: Yes. We can do that.

Comment (9rb) [reply integration-tests/test/test_unbond.py]: @zsluedem has the change proposed by @tgrospic been made?

Comment (zsluedem) [reply integration-tests/test/test_unbond.py]: no

Comment (zsluedem) [reply integration-tests/test/test_unbond.py]: @tgrospic @9rb I will try to use some other ways(meybe websocket) to do that kind of things. I admit right now the way we do is not really elegant.

File: casper/src/main/scala/coop/rchain/casper/BlockCreator.scala [hunk 2]
```diff
@@ -66,8 +79,11 @@ object BlockCreator {
               Tools.rng(invalidBlockHash.toByteArray)
             )
         )
-        deploys          <- extractDeploys(dag, parentMetadatas, maxBlockNumber, expirationThreshold)
-        parents          <- parentMetadatas.toList.traverse(p => ProtoUtil.getBlock(p.blockHash))
+        deploys <- extractDeploys(dag, parentMetadatas, maxBlockNumber, expirationThreshold)
+        parents <- parentMetadatas.toList.traverse(p => ProtoUtil.getBlock(p.blockHash))
+        isActive = parents
+          .traverse(b => isActiveValidator(b, runtimeManager, validatorIdentity))
+          .map(_.foldRight(true)(_ && _))
```
Comment (tgrospic) [inline casper/src/main/scala/coop/rchain/casper/BlockCreator.scala]: Please add some comments about the logic how to decide if it's active validator.

File: rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala [hunk 3]
```diff
@@ -180,18 +181,18 @@ class DebruijnInterpreter[M[_], F[_]](
...
+        case (term, index) =>
+          eval(term)(env, split(index))
+      }
     }
   }
 
+  private def parTraverseSafe[A](xs: Vector[A])(op: A => M[Unit]): M[Unit] =
+    xs.parTraverse(op(_).map(_ => none[Throwable]).handleError(_.some))
+      .map(_.flattenOption)
+      .flatMap(aggregateEvaluatorErrors)
```
Comment (tgrospic) [inline rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala]: These changes don't belong here, please check if you rebased on latest dev.

**Comment by zsluedem** [comment]
@tgrospic How is it now? Please tell me the result because you approved this pr before, i am not sure about how you think now.

**Comment by tgrospic** [comment]
LGTM

File: casper/src/main/scala/coop/rchain/casper/BlockCreator.scala [hunk 4]
```diff
@@ -66,8 +79,19 @@ object BlockCreator {
...
+        // 1. it is possible that you are active in some parents but not active in other parents
+        // 2. all of the parents are in active state
+        // 3. all of the parents are not in active state
+        // Let's just talk about 1 situation because 2 and 3 are easy to judge.
+        // If one validator issue a bonding or unbonding request , it would be possible to end up with
+        // parent in 1 situation. It would be safer for the validator to make sure all parents post state
+        // are in active state.
+        isActive = parents
+          .traverse(b => isActiveValidator(b, runtimeManager, validatorIdentity))
+          .map(_.foldRight(true)(_ && _))
```
Comment (tgrospic) [inline casper/src/main/scala/coop/rchain/casper/BlockCreator.scala]: Checking for all can be little shorter `.map(_.forall(identity))`. :)

File: integration-tests/test/test_bonding_validators.py [hunk 5]
```diff
@@ -39,61 +35,82 @@ def test_heterogenous_validators(command_line_options: CommandLineOptions, rando
...
 
     bonded_validator_map = {
-        BOOTSTRAP_KEY: 200,
-        BONDED_VALIDATOR_KEY: 300
+        BOOTSTRAP_KEY: 10000000,
+        BONDED_VALIDATOR_KEY: 10000000
     }
 
+    epoch_length = 4
+    quarantine_length = 2
```
Comment (Isaac-DeFrain) [inline integration-tests/test/test_bonding_validators.py]: Not sure it's a big deal, but quarantine length should be longer than epoch length, typically it's 5x longer (that factor is not important though).

File: integration-tests/test/test_unbond.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+        VALIDATOR_KEY_1: 20000000,
+        VALIDATOR_KEY_2: 20000000,
+        VALIDATOR_KEY_3: 20000000,
+    }
+
+    # unbond a validator. set the epoch length to 3 and quarantine length to 2
+    # every epoch change the network would re-pick the active validator and
+    # the withdraw process is going to happen after (quarantine_length + epoch_length ) blocks
+    epoch_length = 3
+    quarantine_length = 2
```
Comment (Isaac-DeFrain) [inline integration-tests/test/test_unbond.py]: same as above

File: casper/src/main/scala/coop/rchain/casper/BlockCreator.scala [hunk 7]
```diff
@@ -66,8 +79,19 @@ object BlockCreator {
...
+        deploys <- extractDeploys(dag, parentMetadatas, maxBlockNumber, expirationThreshold)
+        parents <- parentMetadatas.toList.traverse(p => ProtoUtil.getBlock(p.blockHash))
+        // there are 3 situations on judging whether the validator is active
+        // 1. it is possible that you are active in some parents but not active in other parents
+        // 2. all of the parents are in active state
+        // 3. all of the parents are not in active state
+        // Let's just talk about 1 situation because 2 and 3 are easy to judge.
+        // If one validator issue a bonding or unbonding request , it would be possible to end up with
+        // parent in 1 situation. It would be safer for the validator to make sure all parents post state
+        // are in active state.
```
Comment (Isaac-DeFrain) [inline casper/src/main/scala/coop/rchain/casper/BlockCreator.scala:91]: Very nice comments!

File: integration-tests/test/test_bonding_validators.py [hunk 8]
```diff
@@ -39,61 +35,82 @@ def test_heterogenous_validators(command_line_options: CommandLineOptions, rando
...
 
             with bootstrap_connected_peer(context=context, bootstrap=bootstrap_node, name='joining-validator', private_key=JOINING_VALIDATOR_KEY) as joining_validator:
                 wait_for_peers_count_at_least(context, bonded_validator, 2)
+
                 # new joining validator can not propose before bonding
                 with pytest.raises(NonZeroExitCodeError):
                     joining_validator.deploy(contract_path, JOINING_VALIDATOR_KEY)
                     joining_validator.propose()
 
+                # block number 2
```
Comment (Isaac-DeFrain) [inline integration-tests/test/test_bonding_validators.py:68]: What causes the block number to change here?

**Review by Isaac-DeFrain — APPROVED** [review]
Very nice work! Just left a few minor comments/questions.

File: integration-tests/test/test_bonding_validators.py (hunk 8 above)
Comment (zsluedem) [reply integration-tests/test/test_bonding_validators.py:68]: https://github.com/rchain/rchain/pull/2754/files#diff-73712dc027d17012dbd7e2db68f27779L67-R78

**Pull request merged**